# Data Structures in QTL Viewer

The QTL Viewer makes use of R and several different libraries. This document is intended to explain the format of the main `RData` file.

> Please note that some data elements are pre-computed.

The following sections try to explain what each element is required and which is optional, along with the structure.

**mRNA** and **protein** data sets are treated similarly.  **phenotype** data sets have a different structure.  This will hopefully be made clear in the [`dataset.*`](#data_elements) section below.

* [mRNA and protein `dataset`](#mrna_dataset)
* [phenotype `dataset`](#pheno_dataset)


## Elements

The following elements should be in the `RData` file.  

* [`ensembl.version`](#ensembl_version) - number specifying the Ensembl version
* [`genoprobs`](#genoprobs) - the genotype probabilities 
* [`K`](#kinship) - the kinship matrix
* [`map`](#map) - list of one element per chromosome, with the genomic position of each marker
* [`markers`](#markers) - marker names and positions

*Exact case of the variable names is important*

------------

#### `ensembl.version`<a name="ensembl_version"></a> *OPTIONAL*

- **R data type:** numeric

- **Description:** This is the Ensembl information. For example, 90

Please see the documentation at [Ensembl](http://www.ensembl.org/info/website/archives/assembly.html) for build information.

------------

#### `genoprobs`<a name="genoprobs"></a> *REQUIRED*

- **R data type:** [calc_genoprobs](https://github.com/rqtl/qtl2geno)

- **Description:** This is the genotype probabilities. May be produced by [qtl2convert::probs_doqtl_to_qtl2](https://github.com/rqtl/qtl2convert/blob/master/R/probs_doqtl_to_qtl2.R).

Please see the documentation at [R/qtl2geno](https://github.com/rqtl/qtl2geno).

------------

#### `K`<a name="kinship"></a> *REQUIRED*

- **R data type:** [list](http://www.r-tutor.com/r-introduction/list)

- **Description:** A list of kinship matrices, one per chromosome as generated by [qtl2geno::calc_kinship(probs, type = "loco")](https://github.com/rqtl/qtl2geno/blob/master/R/calc_kinship.R).

------------

#### `map`<a name="map"></a> *REQUIRED*

- **R data type:** [list](http://www.r-tutor.com/r-introduction/list)

- **Description:** A list of numeric vectors, one per chromosome. Each vector contains the Mb position of a marker. The list must have names indicating the chromosome. May be produced by [qtl2convert::map_df_to_list](https://github.com/rqtl/qtl2convert/blob/master/R/map_df_to_list.R).

------------

#### `markers`<a name="markers"></a> *REQUIRED*

- **R data type:** [data.frame](http://www.r-tutor.com/r-introduction/data-frame)

- **Description:** marker information 

The following column names must be present:

| column_name | column_type | description |
| ----------- | ---------- | ---------- |    
| `marker` | character | unique name of the marker
| `chr` | character | the chromosome
| `pos` | numeric | position in megabases, (*For example 0.0-200.0*)

`rownames` must also match the names of the `marker` column

------------

## Configurable Elements

The following element is a *special* element and there must be at least per `RData` file.

* [`dataset.*`](#data_elements) - special element

**mRNA**, **protein**, and **phenotype** data sets CAN be configured by `dataset.*` elements.

### `dataset.*`<a name="data_elements"></a>

- **R data type:** [list](http://www.r-tutor.com/r-introduction/list)

- **Description:** annotations for **mRNA** or **protein**
The `*` part should be alpha-numeric text (underscores are allowed) that describes succinctly your data element.  There must be at least 1 `dataset.*` elements per `RData` file.  These elements will allow you to store multiple **mRNA**, **protein**, and **phenotype** data sets. For example, there could be `dataset.mRNA` and `dataset.mRNA2` to store 2 mRNA datasets.  The text for the `*` part should be any text that make sense to you or your dataset.  What is defined in the `dataset.*` element is what is more important.

------------

### mRNA and protein <a name="mrna_dataset"></a>`dataset.*`

The following elements should be in each **mRNA** and **protein** `dataset.*` [list](http://www.r-tutor.com/r-introduction/list)

* [`annots`](#mrna_annots) - annotations of the mrna or protein data
* [`covar`](#mrna_covar) - matrix of covariates data, samples (rows) x covariates (columns)
* [`covar.factors`](#mrna_covar_factors) - specific information about the covars
* [`datatype`](#mrna_datatype) - type of data, either **mRNA** or **protein**
* [`display.name`](#mrna_display_name) - simple display name for the viewer
* [`rankz`](#mrna_expr) - rankZ normalized expression data, samples (rows) x mrna (columns)
* [`norm`](#mrna_norm) - normalized (by method other than rankZ) expression data, samples (rows) x mrna (columns)
* [`lod.peaks`](#mrna_lodpeaks) - lod peaks over a certain threshold
* [`raw`](#mrna_raw) - matrix of raw mrna data (counts out of [EMASE](http://churchill-lab.github.io/emase/))
* [`samples`](#mrna_samples) - sample annotations

------------

#### `annots`<a name="mrna_annots"></a> *REQUIRED*

- **R data type:** [data.frame](http://www.r-tutor.com/r-introduction/data-frame)

- **Description:** annotations for **mRNA** or **protein**

The following column names must be present:

| column_name | column_type | description |
| ----------- | ---------- | ---------- |    
| `protein_id` | character | Ensembl Protein ID **(only if protein, omit column if mRNA)**
| `gene_id` | character | Ensembl Gene ID
| `symbol` | character | the symbol
| `chr` | character | chromosome
| `start` | numeric | start position in megabases, (*For example 0.0-200.0*)
| `end` | numeric | end position in megabases, (*For example 0.0-200.0*)
| `strand` | numeric | -1 for negative, 1 for positive
| `middle` | numeric | middle point in megabases, (*For example 0.0-200.0*)
| `nearest.marker.id` | character | id pointing to [`marker`](#markers)

`rownames` must also match the names of the `id` column

------------

#### `covar`<a name="mrna_covar"></a> *REQUIRED*

- **R data type:** [matrix](http://www.r-tutor.com/r-introduction/matrix)

- **Description:** covariates data, samples (rows) x covariates (columns) as produced by [model.matrix](https://www.rdocumentation.org/packages/stats/versions/3.4.1/topics/model.matrix).

------------

#### `covar.factors`<a name="mrna_covar_factors"></a> *REQUIRED*

- **R data type:** [data.frame](http://www.r-tutor.com/r-introduction/data-frame)

- **Description:** covar information

The following column names must be present:

| column_name | column_type | description |
| ----------- | ---------- | ---------- |    
| `column.name` | character | column name in [`samples`](#mrna_samples)
| `display.name` | character | display name for the viewer
| `int.covar` | character | either `factor`, `numeric` or NA
| `lod.peaks` | character | name of the corresponding element in [`lod.peaks`](#mrna_lodpeak)
| `covar.name` | character | name of the corresponding column in [`covar`](#mrna_covar)

------------

#### `datatype`<a name="mrna_datatype"></a> *REQUIRED*

- **R data type:** character

- **Description:** **mRNA** or **protein**

------------

#### `display.name`<a name="mrna_display_name"></a> *OPTIONAL*

- **R data type:** character

- **Description:** simple display name for the viewer

------------

#### `expr`<a name="mrna_expr"></a> *REQUIRED*

- **R data type:** [matrix](http://www.r-tutor.com/r-introduction/matrix)

- **Description:** expression data, [`samples`](#mrna_samples) (rows) x [`annots`](#mrna_annots) (columns)

------------

#### `norm`<a name="mrna_norm"></a> *OPTIONAL*

- **R data type:** [matrix](http://www.r-tutor.com/r-introduction/matrix)

- **Description:** expression data, [`samples`](#mrna_samples) (rows) x [`annots`](#mrna_annots) (columns)

------------

#### `lod.peaks`<a name="mrna_lodpeaks"></a> *REQUIRED*

- **R data type:** [list](http://www.r-tutor.com/r-introduction/list)

- **Description:** list of peak lod scores above a certain threshold

Each list element is a [data.frame](http://www.r-tutor.com/r-introduction/data.frame). The list elements must be named. The additive scans are named `additive`. Interactive scans (i.e. sex X QTL interaction) are named `sex_int` and must match the `lod.peaks` column in [`covar.factors`](#mrna_covar_factors).

The following column names must be present in each data.frame:

| column_name | column_type | description |
| ----------- | ---------- | ---------- |    
| `annot.id` | character | id pointing to [`annots`](#mrna_annots)
| `marker.id` | character | id pointing to [`marker`](#markers)
| `lod` | numeric | lod score

------------

#### `raw`<a name="mrna_raw"></a> *OPTIONAL*

- **R data type:** [matrix](http://www.r-tutor.com/r-introduction/matrix)

- **Description:** raw expression data, [`samples`](#mrna_samples) (rows) x [`annots`](#mrna_annots) (columns)

------------

#### `samples`<a name="mrna_samples"></a> *REQUIRED*

- **R data type:** [data.frame](http://www.r-tutor.com/r-introduction/data-frame)

- **Description:** sample annotations

The following column names must be present:

| column_name | column_type | description |
| ----------- | ---------- | ---------- |    
| `mouse.id` | character | Mouse ID

`rownames` must also match the names of the `mouse.id` column

Extra columns can be included, and will be rendered in the viewer. For example,

> | mouse.id | Sex | Age | Description |
> | ----------- | ---------- | ---------- | ---------- |
> | DO-001 | F | 12 | some description
> | DO-002 | F | 14 | another description
> | DO-203 | M | 1 | yet another description

------------

### phenotype <a name="pheno_dataset"></a>`dataset.*`

The following elements should be in each **phenotype** `dataset.*` [list](http://www.r-tutor.com/r-introduction/list).

* [`annots`](#pheno_annots) - data dictionary 
* [`covar`](#pheno_covar) - matrix of covariates data, samples (rows) x covariates (columns)
* [`covar.factors`](#pheno_covar_factors) - specific information about the covars
* [`datatype`](#pheno_datatype) - type of data, must be **phenotype**
* [`display.name`](#pheno_display_name) - simple display name for the viewer
* [`lod.peaks`](#pheno_lodpeaks) - lod peaks over a certain threshold
* [`pheno`](#pheno_pheno) -  data, samples (rows) x mrna (columns)
* [`samples`](#pheno_samples) - sample annotations

------------

#### `annots`<a name="pheno_annots"></a> *REQUIRED*

- **R data type:** [data.frame](http://www.r-tutor.com/r-introduction/data-frame)

- **Description:** data dictionary for phenotypes

Formatted as in the [DataDictionaryGuide](https://github.com/churchill-lab/qtl-viewer/blob/master/docs/DataDictionaryGuide.md).

------------

#### `covar`<a name="pheno_covar"></a> *REQUIRED*

- **R data type:** [matrix](http://www.r-tutor.com/r-introduction/matrix)

- **Description:** covariates data, samples (rows) x covariates (columns) as produced by [model.matrix](https://www.rdocumentation.org/packages/stats/versions/3.4.1/topics/model.matrix).

------------

#### `covar.factors`<a name="pheno_covar_factors"></a> *REQUIRED*

- **R data type:** [data.frame](http://www.r-tutor.com/r-introduction/data-frame)

- **Description:** covar information

The following column names must be present:

| column_name | column_type | description |
| ----------- | ---------- | ---------- |    
| `column_name` | character | column name in [`samples`](#pheno_samples)
| `display_name` | character | display name for the viewer
| `int.covar` | character | string, either `factor`, `numeric` or NA
| `lod.peaks` | character | string, name of the corresponding element in [`lod.peaks`](#pheno_lodpeaks)
| `covar.name` | character | name of the corresponding column in [`covar`](#pheno_covar)

------------

#### `datatype`<a name="pheno_datatype"></a> *REQUIRED*

- **R data type:** character

- **Description:** MUST BE **phenotype**

------------

#### `display.name`<a name="pheno_display_name"></a> *OPTIONAL*

- **R data type:** character

- **Description:** simple display name for the viewer

------------

#### `lod.peaks`<a name="pheno_lodpeaks"></a> *REQUIRED*

- **R data type:** [list](http://www.r-tutor.com/r-introduction/list)

- **Description:** list of peak lod scores above a certain threshold

Each list element is a [data.frame](http://www.r-tutor.com/r-introduction/data.frame). The list elements must be named. The additive scans are named `additive`. Interactive scans (i.e. sex X QTL interaction) are named `sex_int` and must match the `lod.peaks` column in [`covar.factors`](#pheno_covar_factors)..

The following column names must be present in each data.frame:

| column_name | column_type | description |
| ----------- | ---------- | ---------- |    
| `annot.id` | character | id pointing to [`annots`](#pheno_annots)
| `marker.id` | character | id pointing to [`marker`](#markers)
| `lod` | numeric | lod score

------------

#### `pheno`<a name="pheno_pheno"></a> *REQUIRED*

- **R data type:** [data.frame](http://www.r-tutor.com/r-introduction/data-frame)

- **Description:** pheno data, [`samples`](#pheno_samples) (rows) x [`annots`](#pheno_annots) (columns)

------------

#### `samples`<a name="pheno_samples"></a> *REQUIRED*

- **R data type:** [data.frame](http://www.r-tutor.com/r-introduction/data-frame)

- **Description:** sample annotations

The following column names must be present:

| column_name | column_type | description |
| ----------- | ---------- | ---------- |    
| `mouse.id` | character | Mouse ID

`rownames` must also match the names of the `mouse.id` column

Extra columns can be included, and will be rendered in the viewer. For example,

> | mouse.id | Sex | Age | Description |
> | ----------- | ---------- | ---------- | ---------- |
> | DO-001 | F | 12 | some description
> | DO-002 | F | 14 | another description
> | DO-203 | M | 1 | yet another description

------------

